File: apps/android/core/model/src/main/kotlin/me/meeshy/sdk/model/Community.kt
package me.meeshy.sdk.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** A community member's user — port of APICommunityUser (CommunityModels.swift). */
@Serializable
data class ApiCommunityUser(
    val id: String,
    val username: String = "",
    val displayName: String? = null,
    val avatar: String? = null,
    val isOnline: Boolean? = null,
)

/** A community member — port of APICommunityMember (CommunityModels.swift). */
@Serializable
data class ApiCommunityMember(
    val id: String,
    val communityId: String = "",
    val userId: String = "",
    val role: String = "member",
    val joinedAt: String? = null,
    val user: ApiCommunityUser? = null,
)

@Serializable
data class ApiCommunityCount(
    val members: Int? = null,
    @SerialName("Conversation") val conversation: Int? = null,
)

/** API community payload — port of APICommunity (CommunityModels.swift). */
@Serializable
data class ApiCommunity(
    val id: String,
    val identifier: String = "",
    val name: String = "",
    val description: String? = null,
    val avatar: String? = null,
    val banner: String? = null,
    val isPrivate: Boolean = false,
    val createdBy: String = "",
    val createdAt: String? = null,
    val updatedAt: String? = null,
    val creator: ApiCommunityUser? = null,
    val members: List<ApiCommunityMember>? = null,
    @SerialName("_count") val count: ApiCommunityCount? = null,
    val memberCount: Int? = null,
    val conversationCount: Int? = null,
)

/** A community search result — port of APICommunitySearchResult (CommunityModels.swift). */
@Serializable
data class ApiCommunitySearchResult(
    val id: String,
    val name: String = "",
    val identifier: String = "",
    val description: String? = null,
    val avatar: String? = null,
    val isPrivate: Boolean = false,
    val memberCount: Int? = null,
    val conversationCount: Int? = null,
    val createdAt: String? = null,
    val creator: ApiCommunityUser? = null,
    val members: List<ApiCommunityMember>? = null,
)

@Serializable
data class CreateCommunityRequest(
    val name: String,
    val identifier: String? = null,
    val description: String? = null,
    val isPrivate: Boolean = true,
)

@Serializable
data class UpdateCommunityRequest(
    val name: String? = null,
    val identifier: String? = null,
    val description: String? = null,
    val isPrivate: Boolean? = null,
    val avatar: String? = null,
    val banner: String? = null,
)

@Serializable
data class InviteMemberRequest(
    val userId: String,
)

@Serializable
data class IdentifierAvailability(
    val available: Boolean = false,
    val identifier: String = "",
)

File: packages/shared/types/api-schemas/community.ts
/**
 * Schémas d’API — communautés et leurs membres.
 *
 * Extrait de `types/api-schemas.ts` par #4635 (découpage du contrat de réponse
 * du dépôt, directive 2026-08-28). Le texte des schémas est INCHANGÉ : seule
 * leur adresse de fichier bouge. `types/api-schemas.ts` reste la FAÇADE qui les
 * ré-exporte, et aucun importeur n’a bougé.
 *
 * @module @meeshy/shared/types/api-schemas/community
 */

import { userMinimalSchema } from './user.js';

// =============================================================================
// COMMUNITY SCHEMAS
// =============================================================================

/**
 * Community schema for API responses
 */
export const communitySchema = {
  type: 'object',
  description: 'Community/group of conversations',
  properties: {
    id: { type: 'string', description: 'Community unique identifier' },
    identifier: { type: 'string', description: 'Human-readable identifier' },
    name: { type: 'string', description: 'Community name' },
    description: { type: 'string', nullable: true, description: 'Community description' },
    avatar: { type: 'string', nullable: true, description: 'Community avatar URL' },
    banner: { type: 'string', nullable: true, description: 'Community banner URL' },
    isPrivate: { type: 'boolean', description: 'Whether community is private' },
    isActive: { type: 'boolean', description: 'Whether community is active' },
    deletedAt: { type: 'string', format: 'date-time', nullable: true, description: 'Deletion timestamp' },
    createdBy: { type: 'string', description: 'Creator user ID' },
    createdAt: { type: 'string', format: 'date-time', description: 'Creation timestamp' },
    updatedAt: { type: 'string', format: 'date-time', description: 'Last update timestamp' },
    creator: { ...userMinimalSchema, description: 'Creator user info' },
    memberCount: { type: 'number', description: 'Number of members' },
    conversationCount: { type: 'number', description: 'Number of conversations' }
  }
} as const;

/**
 * Minimal community schema for lists
 */
export const communityMinimalSchema = {
  type: 'object',
  description: 'Minimal community data for lists',
  properties: {
    id: { type: 'string', description: 'Community ID' },
    identifier: { type: 'string', description: 'Identifier' },
    name: { type: 'string', description: 'Name' },
    avatar: { type: 'string', nullable: true, description: 'Avatar URL' },
    isPrivate: { type: 'boolean', description: 'Is private' },
    memberCount: { type: 'number', description: 'Member count' }
  }
} as const;

/**
 * Community member schema
 */
export const communityMemberSchema = {
  type: 'object',
  description: 'Community membership',
  properties: {
    id: { type: 'string', description: 'Membership unique identifier' },
    communityId: { type: 'string', description: 'Community ID' },
    userId: { type: 'string', description: 'User ID' },
    joinedAt: { type: 'string', format: 'date-time', description: 'Join timestamp' },
    role: {
      type: 'string',
      enum: ['admin', 'moderator', 'member'],
      description: 'Member role'
    },
    isActive: { type: 'boolean', description: 'Whether membership is active' },
    leftAt: { type: 'string', format: 'date-time', nullable: true, description: 'Leave timestamp' },
    user: { ...userMinimalSchema, description: 'User info' }
  }
} as const;

/**
 * Create community request schema
 */
export const createCommunityRequestSchema = {
  type: 'object',
  required: ['name'],
  properties: {
    name: {
      type: 'string',
      minLength: 1,
      maxLength: 100,
      description: 'Community name'
    },
    identifier: {
      type: 'string',
      minLength: 1,
      maxLength: 50,
      pattern: '^[a-zA-Z0-9\\-_]+$',
      description: 'Custom identifier for URLs'
    },
    description: {
      type: 'string',
      maxLength: 500,
      description: 'Community description'
    },
    isPrivate: {
      type: 'boolean',
      default: true,
      description: 'Whether community is private'
    }
  }
} as const;

/**
 * Update community request schema
 */
export const updateCommunityRequestSchema = {
  type: 'object',
  properties: {
    name: { type: 'string', minLength: 1, maxLength: 100 },
    description: { type: 'string', maxLength: 500 },
    avatar: { type: 'string', format: 'uri' },
    banner: { type: 'string', format: 'uri' },
    isPrivate: { type: 'boolean' }
  }
} as const;
